### PR TITLE
Scan QRcode

### PR DESCRIPTION
**Pull request by makoto**
## Changes

- Move search to use local state to global state
- Add QRCode mutation and query
- Display the first 4~5 ETH address if you are the contract admin
- Shows "Scan QR code" only via supported browser
- Scans QRcode and search by the address

## Demo video

https://www.youtube.com/watch?v=9IRkOWOvC5k&feature=youtu.be

(Tested on Cipher and Status.im, Coinbase Wallet is not supported)

## How to test

1. Change `src/config/env.json` to point to rinkeby

```
{
  "DEPLOYER_CONTRACT_ADDRESS": "0xfe4d296f7b7e5c0065d5df7865075d93d88415e1",
  "API_URL": "https://rinkeby.api.kickback.events",
  "NUM_CONFIRMATIONS": 1
}
```

2. Install https://ngrok.com and proxy your local node

```
ngrok http 3000
```

NOTE: Do not hit the URL end point too often as it has fairly low rate limiting 

3. Open the proxy url from mobile wallet (Status, Ciphyer, Coinbase wallet)

4. Create test event (or create it via metamask using the same seed as the one you have on your mobile wallet)

5. RSVP from an account

6. Open the RSVPed account via Etherscan (which has QR code section)

7. Click "Scan QR code" and scan the QRCode of #6.

8. It should fill in the scanned QR code address into search box.

### CODE REVIEW

File: src/GlobalState.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ class Provider extends Component {
     currentModal: null,
     auth: LocalStorage.getItem(AUTH) || {},
     networkState: {},
+    searchTerm: ''
```
Comment (hiddentao) [inline src/GlobalState.js]: Why have we shifted search into global state? Global state is only for stuff that applies across the entire app. The search should be contained within the single event page component tree

File: src/api/resolvers/qrCodeResolvers.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      }
+      return {
+        supported:supported,
+        __typename: 'QRCodeSupported'
+      }
+    }
+  },
+  
+  Mutation: {
+    async scanQRCode() {
```
Comment (hiddentao) [inline src/api/resolvers/qrCodeResolvers.js]: This should not be a mutation since it doesn't change any data.

File: src/graphql/mutations.js [hunk 3]
```diff
@@ -120,3 +120,11 @@ export const WithdrawPayout = gql`
     withdrawPayout(address: $address) @client
   }
 `
+
+export const ScanQRCode = gql`
+mutation scanQRCode($address: String) {
```
Comment (hiddentao) [inline src/graphql/mutations.js]: Should not be a mutation since it doesn't update data

File: src/graphql/queries.js [hunk 4]
```diff
@@ -47,3 +47,11 @@ export const ReverseRecordQuery = gql`
     }
   }
 `
+
+export const QRSupportedQuery = gql`
+  query scanQRCodeSupported($address: String) {
```
Comment (hiddentao) [inline src/graphql/queries.js]:
```suggestion
  query scanQRCodeSupported($address: String) {
    supported: scanQRCodeSupported(address: $address) @client
  }
```

File: src/api/resolvers/qrCodeResolvers.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import getWeb3 from '../web3'
+export const defaults = {}
+
+const resolvers = {
+  Query: {
+    async scanQRCodeSupported() {
```
Comment (hiddentao) [inline src/api/resolvers/qrCodeResolvers.js]:
```suggestion
    async scanQRCodeSupported() {
      return !!web3.currentProvider.scanQRCode
    }
```

File: src/components/SingleEvent/Participant.js [hunk 6]
```diff
@@ -103,9 +114,12 @@ export class Participant extends Component {
              <TwitterAvatar user={user} />
             <ParticipantId>
               <ParticipantUsername>{user.username}</ParticipantUsername>
-              {user.realName ? (
+              {amAdmin && user.realName ? (
                 <ParticipantRealName>{user.realName}</ParticipantRealName>
               ) : null}
+              {amAdmin ? (
```
Comment (hiddentao) [inline src/components/SingleEvent/Participant.js]: What does this do to the mark attended button display?

ping @jefflau

File: src/graphql/mutations.js (hunk 3 above)
Comment (hiddentao) [reply src/graphql/mutations.js]: Also, this does not need to take any input variables, please remove `$address`

File: src/components/SingleEvent/EventParticipants.js [hunk 7]
```diff
@@ -47,13 +56,41 @@ class EventParticipants extends Component {
...
+                              onClick={ (() => {
+                              scanQRCode().then((result)=>{
+                                setSearchTerm(result.data.scanQRCode.address)
+                              })
+                            }) }>Scan QRCode</Button>
+                          </QRCodeContainer>
+                        )}  
+                      </SafeMutation>          
+                    )
+                  } else {
```
Comment (hiddentao) [inline src/components/SingleEvent/EventParticipants.js]: I would rather we display nothing if QR code scanning is not supported.

File: src/components/SingleEvent/EventParticipants.js [hunk 8]
```diff
@@ -47,13 +56,41 @@ class EventParticipants extends Component {
...
+                variables={{ address: '1' }}
+              >
+                {result => {
+                  if (result.data.scanQRCodeSupported && result.data.scanQRCodeSupported.supported) {
+                    return(
+                      <SafeMutation mutation={ScanQRCode}>
+                        {scanQRCode => (
+                          <QRCodeContainer>
+                            <Button 
+                              onClick={ (() => {
```
Comment (hiddentao) [inline src/components/SingleEvent/EventParticipants.js]:
```suggestion
                              onClick={this._scan({ scanQRCode, setSearchTerm })
```

Then in `EventParticipants` class:

```
_scan = ({ scanQRCode, setSearchTerm }) => async () => {
  const result =await scanQRCode()
  const code = _.get(result, 'data.scanQRCode.address')
  if (code) {
    setSearchTerm(code)
  }
}
```

Note that `_` = `lodash` module

File: src/api/resolvers/qrCodeResolvers.js (hunk 2 above)
Comment (makoto) [reply src/api/resolvers/qrCodeResolvers.js]: Moved to query

File: src/api/resolvers/qrCodeResolvers.js (hunk 5 above)
Comment (makoto) [reply src/api/resolvers/qrCodeResolvers.js]: If I do this, how does the following query changes? 

```
export const QRSupportedQuery = gql`
  query scanQRCodeSupported {
    scanQRCodeSupported @client {
      supported
    }
  }
`
```

File: src/components/SingleEvent/EventParticipants.js (hunk 7 above)
Comment (makoto) [reply src/components/SingleEvent/EventParticipants.js]: Fixed

File: src/GlobalState.js (hunk 1 above)
Comment (makoto) [reply src/GlobalState.js]: This line was duplicated into 3 different components so I thought it's better to move to global state
```
  state = {
    search: ''
  }
   handleSearch = event => {
    this.setState({
      search: event.target.value
    })
  }
```

Comment (hiddentao) [reply src/GlobalState.js]: I think there is redundant code in the components, that's why it looks like things are being duplicated all over. I would say we should search to be contained with `EventParticipants`.

File: src/api/resolvers/qrCodeResolvers.js (hunk 5 above)
Comment (hiddentao) [reply src/api/resolvers/qrCodeResolvers.js]:
```suggestion
export const QRSupportedQuery = gql`
  query scanQRCodeSupported {
    supported: scanQRCodeSupported @client
  }
```

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Comment by noblocknoparty-devs** [comment]
PR deployed at: https://pr22-app-noblocknoparty.surge.sh

**Pull request merged**